README
W3C-LinkChecker
===============

This distribution contains the W3C Link Checker.

The link checker can be run as a CGI script in a web server as well as
on the command line.  The CGI version provides a HTML interface as
seen at <http://validator.w3.org/checklink>.

To install the distribution for command line use:

   perl Makefile.PL
   make
   make test
   make install # as root

To install the CGI version, in addition to the above, copy the
bin/checklink script into a location in your web server from where
execution of CGI scripts is allowed, and make sure that the web server
user has execute permissions to the script.  The CGI directory is
typically named "cgi-bin" somewhere under your web server root
directory.

For more information, please consult the POD documentation in the
checklink.pod file, typically (in the directory where you unpacked the
source):

  perldoc ./bin/checklink.pod

...as well as the HTML documentation in docs/checklink.html.


COPYRIGHT AND LICENCE

Written by the following people for the W3C:
- Hugo Haas <hugo@w3.org>
- Ville Skyttä <ville.skytta@iki.fi>
- W3C QA-dev Team <public-qa-dev@w3.org>

Copyright (C) 1994-2011 World Wide Web Consortium, (Massachusetts
Institute of Technology, European Research Consortium for Informatics
and Mathematics, Keio University). All Rights Reserved. This work is
distributed under the W3C(R) Software License [1] in the hope that it
will be useful, but WITHOUT ANY WARRANTY; without even the implied
warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

[1] http://www.w3.org/Consortium/Legal/copyright-software


NEWS
This document contains information about high level changes between
Link Checker releases.

Version 4.81 - 2011-10-16
- Work around some <base> related problems (#12720, rt.cpan.org#54361).
- Eliminate some warnings (emitted by code, not from results).

Version 4.8 - 2011-04-02
- Avoid some robot delays by improving the order in which links are checked.
- Avoid some unnecessary HEAD requests in recursive mode.
- Clarify output wrt. links that have already been checked.
- Make connection cache size configurable, and increase the default to 2.
- Move JavaScript to an external file.
- Check applet and object archive links.

Version 4.7 - 2011-03-17
- Support for IRI.
- Support for more HTML5 links.
- Decode query string parameters as UTF-8.
- Decode command line arguments according to locale.
- New dependencies: Encode-Locale (command line mode only).
- Updated dependencies: libwww-perl >= 5.833, URI >= 1.53.

Version 4.6 - 2010-05-01
- Support for checking links in CSS.
- Results UI improvements, added "progress bar".
- Support for larger variety of character and content encodings.
- Support for HTTP responses with > 4kB header lines.
- Additional output suppress options in command line mode.
- Improved heuristics when passed non-absolute URLs.
- Support for cookies (command line only for now).
- More "false positive" failure avoidance efforts for "make test".
- The set of forbidden protocols is now configurable.
- New dependencies: CSS-DOM >= 0.09.
- Updated dependencies: Perl >= 5.8.

Version 4.5 - 2009-03-30
- Removed W3C trademarked icons from distribution tarball.
- Avoid "false positive" failures from "make test" in certain setups.
- Make quiet command line mode quieter.
- Lowered default timeout to 30 seconds.

Version 4.4 - 2009-02-12
- checking more elements and attributes, such as BLOCKQUOTE cite="", BODY
  background="", EMBED, etc
- Changes in the UI to make it match other validators more closely
- in HTML/cgi output, using javascript to show checklink status as it happens
- added support for HTML5 links
- softer wording for broken link results
- Add non-robot developer mode
- many bug fixes and code cleanup

Version 4.3 - 2006-10-22
- Various minor improvements to result output, both in text and HTML modes.
- Fix --quiet and checking multiple documents to match documentation.
- Eliminate various warnings (emitted by code, not from results).
- Documentation improvements.

Version 4.2.1 - 2005-05-15
- Include documentation of the reorganized access keys.

Version 4.2 - 2005-04-27
- Access key reorganization, making them less likely to conflict with
  browsers' "native" key bindings.
- Redirects are now checked for private IP addresses too.

Version 4.1 - 2004-11-24
- Added workarounds against browser timeouts in "summary only" mode.
- Improved caching and reuse of fetched /robots.txt information.
- Fixed a bug where a complete protocol response (including headers)
  was passed to the HTML parser, which caused unexpected behaviour.
- Minor user interface and installation related improvements.


Makefile.PL
use 5.008;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME         => 'W3C::LinkChecker',
    ABSTRACT     => 'W3C Link Checker',
    AUTHOR       => 'W3C QA-dev Team <public-qa-dev@w3.org>',
    LICENSE      => 'open_source',
    VERSION_FROM => 'bin/checklink',
    PREREQ_PM    => {

        # Hard dependencies:
        CSS::DOM            => 0.09,
        CSS::DOM::Constants => 0,
        CSS::DOM::Style     => 0,
        CSS::DOM::Util      => 0,
        Encode              => 0,
        HTML::Entities      => 0,
        HTML::Parser        => "3.40",
        HTTP::Headers::Util => 0,
        HTTP::Message       => 5.827,
        HTTP::Request       => 0,
        HTTP::Response      => "1.50",
        LWP::RobotUA        => 1.19,
        LWP::UserAgent      => 0,
        Net::HTTP::Methods  => 5.833,
        Time::HiRes         => 0,
        URI                 => 1.53,
        URI::Escape         => 0,

        # Optional, but required if using a config file:
        Config::General => 2.06,

        # Optional, but required if private IPs are disallowed:
        Net::hostent => 0,
        Net::IP      => 0,
        Socket       => 0,

        # Optional, but required in command line mode:
        File::Spec   => 0,
        Getopt::Long => 2.17,
        Text::Wrap   => 0,
        URI::file    => 0,

        # Optional, used for password input in command line mode:
        Term::ReadKey => 2.00,

        # Optional, used for guessing language in command line mode:
        Locale::Country  => 0,
        Locale::Language => 0,

        # Optional, used when decoding arguments in command line mode:
        Encode::Locale => 0,

        # Optional, but required in CGI mode:
        CGI         => 0,
        CGI::Carp   => 0,
        CGI::Cookie => 0,

        # Optional, required if using cookies:
        HTTP::Cookies => 0,

        # Required for the test suite:
        File::Spec => 0,
        Test::More => 0,
    },
    PM => {'lib/W3C/LinkChecker.pm' => '$(INST_LIB)/W3C/LinkChecker.pm'},
    EXE_FILES => ['bin/checklink'],
    MAN1PODS =>
        {'bin/checklink.pod' => '$(INST_MAN1DIR)/checklink.$(MAN1EXT)',},
    META_MERGE => {
        resources => {
            homepage   => 'http://validator.w3.org/checklink',
            bugtracker => 'http://www.w3.org/Bugs/Public/',
            repository => 'http://dvcs.w3.org/hg/link-checker/',
            MailingList =>
                'http://lists.w3.org/Archives/Public/www-validator/',
        },
    },
    depend => {distdir  => 'lib/W3C/LinkChecker.pm'},
    dist   => {TARFLAGS => '--owner=0 --group=0 -cvf'},
    clean  => {FILES    => 'Makefile.PL.bak bin/checklink.bak'},
);

sub MY::postamble
{
    return <<'MAKE_FRAG';
lib/W3C/LinkChecker.pm: Makefile.PL bin/checklink
	$(MKPATH) lib/W3C
	$(ECHO) "# Dummy module for CPAN indexing purposes."  > $@
	$(ECHO) "package $(NAME);"                           >> $@
	$(ECHO) "use strict;"                                >> $@
	$(ECHO) "use vars qw(\$$VERSION);"                   >> $@
	$(ECHO) "\$$VERSION = \"$(VERSION)\";"               >> $@
	$(ECHO) "1;"                                         >> $@

PERLTIDY = perltidy --profile=etc/perltidyrc --backup-and-modify-in-place \
	--backup-file-extension=/

perltidy:
	@for file in Makefile.PL bin/checklink ; do \
		echo "$(PERLTIDY) $$file" ; \
		$(PERLTIDY) $$file ; \
	done
MAKE_FRAG
}
